extmod/machine_i2c.c

```diff
@@ -75,21 +76,23 @@ STATIC int mp_hal_i2c_sda_read(machine_i2c_obj_t *self) {
     return mp_hal_pin_read(self->sda);
 }
 
-STATIC void mp_hal_i2c_start(machine_i2c_obj_t *self) {
+STATIC int mp_hal_i2c_start(machine_i2c_obj_t *self) {
     mp_hal_i2c_sda_release(self);
     mp_hal_i2c_delay(self);
-    mp_hal_i2c_scl_release(self);
+    int ret = mp_hal_i2c_scl_release(self);
     mp_hal_i2c_sda_low(self);
     mp_hal_i2c_delay(self);
+    return ret;
 }
```

Comment:
If there is a bus error, don't you think the I2C driver should release both SCL and SDA lines before returning?  In this function SCL is released, and so is SDA, right after the call to mp_hal_i2c_scl_release, so you can simply return at that point if the call fails.